### PR TITLE
Fix key dtype in GPTJ and CodeGen

### PR DESCRIPTION
**Pull request by fxmarty**
As per title.

The issue is that 

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

model_id = "hf-internal-testing/tiny-random-GPTJForCausalLM"
model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.float16)
tokenizer = AutoTokenizer.from_pretrained(model_id)

inp = tokenizer("this is me", return_tensors="pt").to("cuda")
model = model.to("cuda")

model = model.eval()
with torch.no_grad():
    res = model(**inp, use_cache=True)

print("res.past_key_values[0][0].dtype", res.past_key_values[0][0].dtype)
print("res.past_key_values[0][1].dtype", res.past_key_values[0][1].dtype)
```
gives
```
res.past_key_values[0][0].dtype torch.float32
res.past_key_values[0][1].dtype torch.float16
```

the reason being that `sin` and `cos` are not properly casted to the correct dtype contrarily to the implementation in e.g. llama.

### CODE REVIEW

File: src/transformers/models/gptj/modeling_gptj.py [hunk 1]
```diff
@@ -55,9 +55,9 @@
 ]
 
 
-def create_sinusoidal_positions(num_pos: int, dim: int) -> torch.Tensor:
+def create_sinusoidal_positions(num_pos: int, dim: int, dtype: torch.dtype) -> torch.Tensor:
     inv_freq = 1.0 / (10000 ** (torch.arange(0, dim, 2) / dim))
-    sinusoid_inp = torch.einsum("i , j -> i j", torch.arange(num_pos, dtype=torch.float), inv_freq).float()
+    sinusoid_inp = torch.einsum("i , j -> i j", torch.arange(num_pos, dtype=torch.float), inv_freq).to(dtype)
```
Comment (ArthurZucker) [inline src/transformers/models/gptj/modeling_gptj.py:60]: I think I specifically kept this for BC actually 😓 The original model does the same and the computation is not the same no?

Comment (fxmarty) [reply src/transformers/models/gptj/modeling_gptj.py:60]: Will double check and keep you up

**Comment by fxmarty** [comment]
Thank you @ArthurZucker for the reader. Indeed in the original codebase of GPTJ and codegen `sincos` is kept as fp32, and the ROPE computation as well as `_attn` computation is done with the `key` always in fp32. The looks very much like a bug in the original implementation of those models.

Anyway, in the interest of keeping the compute similar to the original implementation, I simply cast the key cache to the correct dtype.

**Review by ArthurZucker — APPROVED** [review]
LGTM but can you make extra checks regarding the logits. Wondering if the mixed precision ( float32 x float16 vs now float16 x float16) will be absolutely similar or not

**Comment by fxmarty** [comment]
It is still float16 x float32. We simply store the KV cache as float16. This `key = torch.cat((past_key, key), dim=-2)` basically casts everything to float32 (while `past_key` is fp16, `key` is fp32)

**Comment by ArthurZucker** [comment]
it is for llama 🙄
